### PR TITLE
sql: support array append and equality

### PR DESCRIPTION
**Pull request by justinj**
This commit adds support for the array append operator (||) for both
elements and arrays, along with array equality testing.

### CODE REVIEW

**Comment by knz** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/array, line 634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSYiiIKSLfakNk01wf:-KqSYiiIKSLfakNk01wg:biidfyb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/logictest/testdata/logic_test/array#L634)):*
> ```
> {1,2,3,4}
> 
> # This is a departure from Postgres' behaviour.
> ```

Please expand the comment to motivate the departure.

It seems to me that this is a quirk of our type inference. 
I read this as follows: for an array type A with element type T,
- postgres infers A for x in A || x
- whereas we infer T for x in A || x

If we had the same inference, then an untyped NULL in that position would infer as type A, and then your array-array overload would kick in and do the same as pg. 

---

*[pkg/sql/parser/eval.go, line 211 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSZaEYi-zFbqZyh2hv:-KqSZaEYi-zFbqZyh2hw:b-uqq6w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/parser/eval.go#L211)):*
> ```Go
> 			nullableArgs: true,
> 			fn: func(_ *EvalContext, left Datum, right Datum) (Datum, error) {
> 				result := NewDArray(typ)
> ```

Can you double check what is expected if both arguments are NULL.

---

*[pkg/sql/parser/eval.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSZfgOnbVeclXQJuj6:-KqSZfgOnbVeclXQJuj7:b-rqxvo7) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/parser/eval.go#L232)):*
> ```Go
> 			nullableArgs: true,
> 			fn: func(_ *EvalContext, left Datum, right Datum) (Datum, error) {
> 				result := NewDArray(typ)
> ```

ditto.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by justinj** [comment]
Review status: 3 of 4 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/logictest/testdata/logic_test/array, line 634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSYiiIKSLfakNk01wf:-KqTHOmfBJ9ZSPC0ZdUI:b-plso0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/logictest/testdata/logic_test/array#L634)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please expand the comment to motivate the departure.

It seems to me that this is a quirk of our type inference. 
I read this as follows: for an array type A with element type T,
- postgres infers A for x in A || x
- whereas we infer T for x in A || x

If we had the same inference, then an untyped NULL in that position would infer as type A, and then your array-array overload would kick in and do the same as pg. 
</blockquote></details>

It's actually because during *typechecking* we propagate `TypeNull` values. See [here](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/parser/type_check.go#L129-L133). If we remove that block then the overload checker doesn't know how to resolve cases where there's a NULL argument. We could maybe make the change and also mark one of these overloads as preferred? I'm not sure I know our type system well enough to know the implications of making such a change, so maybe we should discuss.

---

*[pkg/sql/parser/eval.go, line 211 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSZaEYi-zFbqZyh2hv:-KqTGY0kr1ezR7XTKcjt:be6721h) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/parser/eval.go#L211)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you double check what is expected if both arguments are NULL.
</blockquote></details>

The array is treated as an empty array, and the value is `NULL`, which matches Postgres (I added a test):
```
SELECT NULL::INT[] || NULL::INT, NULL::INT || NULL::INT[];
 ?column? | ?column?
----------+----------
 {NULL}   | {NULL}
(1 row)
```

---

*[pkg/sql/parser/eval.go, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSZfgOnbVeclXQJuj6:-KqTHCR5XC2cqk7yStSS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/parser/eval.go#L232)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

ditto.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by knz** [comment]
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/array, line 634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSYiiIKSLfakNk01wf:-KqTPRrU4GquuA0BGcGt:b9qc336) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/logictest/testdata/logic_test/array#L634)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

It's actually because during *typechecking* we propagate `TypeNull` values. See [here](https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/parser/type_check.go#L129-L133). If we remove that block then the overload checker doesn't know how to resolve cases where there's a NULL argument. We could maybe make the change and also mark one of these overloads as preferred? I'm not sure I know our type system well enough to know the implications of making such a change, so maybe we should discuss.
</blockquote></details>

Please check what I'm about to write, but it seems to me that as soon as we started to support overloads that accept null arguments, then this specific code you linked to needs to be altered.

It seems that what is wanted is: if any of the overloads takes nulls as argument, then do not propagate typenull.

Does this make sense?

---

*[pkg/sql/logictest/testdata/logic_test/array, line 625 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqTPnQH-rIG63MxfUdB:-KqTPnQH-rIG63MxfUdC:bb6mt9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/6060e5b27edde40ac2c5262af09aa023d4538ff2/pkg/sql/logictest/testdata/logic_test/array#L625)):*
> ```
> 
> query TT
> SELECT NULL::INT || NULL::INT[], NULL::INT[] || NULL::INT
> ```

:+1:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by jordanlewis** [comment]
:lgtm: mod @knz's question

---

Reviewed 3 of 4 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/sql/logictest/testdata/logic_test/array, line 634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSYiiIKSLfakNk01wf:-KqTj8tYDAOhUvstASyp:bpvtk5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/logictest/testdata/logic_test/array#L634)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please check what I'm about to write, but it seems to me that as soon as we started to support overloads that accept null arguments, then this specific code you linked to needs to be altered.

It seems that what is wanted is: if any of the overloads takes nulls as argument, then do not propagate typenull.

Does this make sense?
</blockquote></details>

I think I agree with Raphael here. ANd you've already made an analogous change to the normalize code, right?

---

*[pkg/sql/parser/normalize.go, line 131 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqTiez5GcuIbc0K1Eqo:-KqTiez6Je7Mdj-hH99i:b-ftw299) ([raw file](https://github.com/cockroachdb/cockroach/blob/6060e5b27edde40ac2c5262af09aa023d4538ff2/pkg/sql/parser/normalize.go#L131)):*
> ```Go
> 	expectedType := expr.ResolvedType()
> 
> 	if !expr.fn.nullableArgs && (left == DNull || right == DNull) {
> ```

This seems important - it concerns me that this wasn't caught by any tests after the nullableArgs PR. Are we missing some tests for this in general?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-KqTitAVdbN8_q2gXplr:bio657a)*

**Comment by justinj** [comment]
Review status: 3 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/logictest/testdata/logic_test/array, line 634 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqSYiiIKSLfakNk01wf:-KqTtUoiuR9Y54ygFz6U:b-vztwig) ([raw file](https://github.com/cockroachdb/cockroach/blob/f2936b039e8be7f1aefef7e70f2b0cdda479e07e/pkg/sql/logictest/testdata/logic_test/array#L634)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think I agree with Raphael here. ANd you've already made an analogous change to the normalize code, right?
</blockquote></details>

OK, PTAL, we just error now. I think in doing so I addressed what might have actually been a bug in the original overload PR, so it only shows the valid candidates. Now running `ARRAY[1,2,3] || NULL` gives
```
pq: ambiguous binary operator: <int[]> || <NULL>, candidates are:
||(int[], int) -> int[]
||(int[], int[]) -> int[]
```
Further changes will be needed for non-BinOp builtins.

---

*[pkg/sql/parser/normalize.go, line 131 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqTiez5GcuIbc0K1Eqo:-KqTjnLdAvQld-BQo32p:b-oi30r6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6060e5b27edde40ac2c5262af09aa023d4538ff2/pkg/sql/parser/normalize.go#L131)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This seems important - it concerns me that this wasn't caught by any tests after the nullableArgs PR. Are we missing some tests for this in general?
</blockquote></details>

The original overloads PR only addressed functions and not `BinaryExpr`s, and in theory we could have made this change then, but we wouldn't have a way to test it (or even observe it) without introducing a `BinaryExpr` that accepts nullableArgs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by jordanlewis** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/normalize.go, line 131 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqTiez5GcuIbc0K1Eqo:-KqUVZ5BRgANpWtn_zvJ:b-n0zrpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6060e5b27edde40ac2c5262af09aa023d4538ff2/pkg/sql/parser/normalize.go#L131)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

The original overloads PR only addressed functions and not `BinaryExpr`s, and in theory we could have made this change then, but we wouldn't have a way to test it (or even observe it) without introducing a `BinaryExpr` that accepts nullableArgs.
</blockquote></details>

Gotcha, cool

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by knz** [comment]
:LGTM: but you need to update the test before merging.

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-KqV5EOGxmVx80rIdC5q:b-fhs2bc)*

**Comment by nvanbenschoten** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/logictest/testdata/logic_test/array, line 738 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqYVR-FhRnYPsPT7uAY:-KqYVR-FhRnYPsPT7uAZ:b9uqoqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/14f9ec743f1f87c1400faeb50f8b2fdcfbfeb293/pkg/sql/logictest/testdata/logic_test/array#L738)):*
> ```
> # NULL), but it's how Postgres behaves.
> query B
> SELECT ARRAY[1,2,NULL] = ARRAY[1,2,3]
> ```

This is strange because `SELECT (1,2,NULL) = (1,2,3)` gives `NULL`, but I guess it's what Postgres gives.

---

*[pkg/sql/parser/overload.go, line 476 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqYZYROQobaVyTFWPXL:-KqYZYROQobaVyTFWPXM:b81qcvb) ([raw file](https://github.com/cockroachdb/cockroach/blob/14f9ec743f1f87c1400faeb50f8b2fdcfbfeb293/pkg/sql/parser/overload.go#L476)):*
> ```Go
> 	}
> 
> 	possibleOverloads := make([]overloadImpl, len(s.overloadIdxs))
> ```

What's going on here? Why did this change?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-KqYZepp6hUy-YhQtndy:bnfp4nl)*

**Comment by knz** [comment]
ugh I still want an answer on the $1 || string situation at least

**Comment by justinj** [comment]
Nathan observed that in ambiguous cases like this it seems that Postgres assumes that the `NULL` arg's type matches the type of the other arg (so `NULL||string` infers the `NULL` to be `string`, and `NULL||string[]` infers it to be `string[]`), which I think seems reasonable, so I'm going to try doing that and see where that takes us.

**Comment by justinj** [comment]
OK, PTAL. A little hairy but I think this gives us the behaviour we want.

---

Review status: 4 of 6 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/parser/overload.go, line 476 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqYZYROQobaVyTFWPXL:-KqcdcHU1sUPkcWwm9qQ:bofvf1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/14f9ec743f1f87c1400faeb50f8b2fdcfbfeb293/pkg/sql/parser/overload.go#L476)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What's going on here? Why did this change?
</blockquote></details>

I think this might be an oversight of Peter's refactoring, this was always returning every overload, rather than the subset that were candidates. There's an extra allocation now but I think it's no big deal since it's only in cases that don't resolve.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-KqcdqHZy5MNhWWjCiIh:b-hc8pem)*

**Comment by nvanbenschoten** [comment]
Reviewed 1 of 2 files at r4.
Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/overload.go, line 469 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpK_kzOkSwxQAbiTd:-KqdpK_kzOkSwxQAbiTe:b-9p3lz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L469)):*
> ```Go
> 	if len(s.exprs) == 2 {
> 		if ok, fns, err := filterAttempt(ctx, &s, func() {
> 			left, err := s.exprs[0].TypeCheck(ctx, TypeAny)
> ```

Aren't these already available in `s.typedExprs[0]` and `s.typedExprs[1]`?

---

*[pkg/sql/parser/overload.go, line 479 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpzMQJ-9UPMvabz6Q:-KqdpzMRNFNSFmZRJv8e:b-1vmkfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L479)):*
> ```Go
> 			leftType := left.ResolvedType()
> 			rightType := right.ResolvedType()
> 			if leftType != TypeNull || rightType != TypeNull {
> ```

We want this to be an XOR, right? Something like `(leftType == TypeNull) != (rightType == TypeNull)`. Otherwise this isn't trying anything new.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by justinj** [comment]
Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/overload.go, line 469 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpK_kzOkSwxQAbiTd:-Kqdqqcy5rpbEeuScJg7:b3uedbd) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L469)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Aren't these already available in `s.typedExprs[0]` and `s.typedExprs[1]`?
</blockquote></details>

My understanding is that those are only ever filled in with respect to a particular overload (and it's really only used as a scratch slice to return the types), see `checkReturn`.

---

*[pkg/sql/parser/overload.go, line 479 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpzMQJ-9UPMvabz6Q:-KqduZNHf6MZRsAkAeSz:bng12lm) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L479)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We want this to be an XOR, right? Something like `(leftType == TypeNull) != (rightType == TypeNull)`. Otherwise this isn't trying anything new.
</blockquote></details>

Ah yes, this could result in some repeat work if they're both non-NULL, you're right.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by nvanbenschoten** [comment]
Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/overload.go, line 469 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpK_kzOkSwxQAbiTd:-KqeBS9utIjHS8HYtk0V:bghm5nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L469)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

My understanding is that those are only ever filled in with respect to a particular overload (and it's really only used as a scratch slice to return the types), see `checkReturn`.
</blockquote></details>

For constants and placeholders they will be filled in later in `checkReturn`, but for `resolvableIdxs`, they should [already be](https://github.com/cockroachdb/cockroach/blob/2e76b6e567311330041c97e2f564ece8485feefd/pkg/sql/parser/overload.go#L326) type checked. Can we conditionally type check based on whether `s.typedExprs[0 or 1]` is nil?

---

*[pkg/sql/parser/overload.go, line 479 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpzMQJ-9UPMvabz6Q:-KqeD1MxU0pKYVgYJfcv:by27a2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L479)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Ah yes, this could result in some repeat work if they're both non-NULL, you're right.
</blockquote></details>

You can replace with `leftIsNull != rightIsNull`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by justinj** [comment]
Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/overload.go, line 469 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpK_kzOkSwxQAbiTd:-KqeM5NJYLfkVRJsimUa:b-fvafj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L469)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

For constants and placeholders they will be filled in later in `checkReturn`, but for `resolvableIdxs`, they should [already be](https://github.com/cockroachdb/cockroach/blob/2e76b6e567311330041c97e2f564ece8485feefd/pkg/sql/parser/overload.go#L326) type checked. Can we conditionally type check based on whether `s.typedExprs[0 or 1]` is nil?
</blockquote></details>

good point, ok!

---

*[pkg/sql/parser/overload.go, line 479 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqdpzMQJ-9UPMvabz6Q:-KqeG4nzRImXyjBPeU3m:b-um3vue) ([raw file](https://github.com/cockroachdb/cockroach/blob/55e6e83dd20ad953bacccc0930637383122b1ecd/pkg/sql/parser/overload.go#L479)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You can replace with `leftIsNull != rightIsNull`
</blockquote></details>

personally I find that a bit hard to parse, so unless you feel strongly I'd rather leave it as it is at the expense of some brevity.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by nvanbenschoten** [comment]
:lgtm:

---

Review status: 5 of 6 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-Kqg0rF_DH137RIbr6h3:bnfp4nl)*

**Comment by knz** [comment]
One more thing...

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/overload.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqggLSsbpa7xDWRb5vT:-KqggLSsbpa7xDWRb5vU:bqzsnnn) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3a3e3fd01da018df742cb0793b04623381a5b2/pkg/sql/parser/overload.go#L467)):*
> ```Go
> 	// other argument. This is used to differentiate the behaviour of
> 	// STRING[] || NULL and STRING || NULL.
> 	if len(s.exprs) == 2 {
> ```

This condition is not doing what the comment is saying.
I think you are right and this behavior should be limited to binary expressions, not every function that happens to have 2 arguments.
So maybe you need to extend the argument list of `typeCheckOverloadedExprs` with a boolean that indicates whether you are in the context of a binary expression, and use that here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-KqggcH722WOu2UD8Rho:b-ldmgt0)*

**Comment by justinj** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/overload.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqggLSsbpa7xDWRb5vT:-KqhmNVBm1nZtjkDNGCv:b-c3gy7r) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3a3e3fd01da018df742cb0793b04623381a5b2/pkg/sql/parser/overload.go#L467)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This condition is not doing what the comment is saying.
I think you are right and this behavior should be limited to binary expressions, not every function that happens to have 2 arguments.
So maybe you need to extend the argument list of `typeCheckOverloadedExprs` with a boolean that indicates whether you are in the context of a binary expression, and use that here.
</blockquote></details>

Would you be OK with doing a typecheck on `overloadImpl`s to see if they're a `BinOp` in the filter here? I'm sort of hesitant to add more inputs to this function that already (necessarily) has a lot going on.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by knz** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/overload.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqggLSsbpa7xDWRb5vT:-Kqm-B-SEvod7zphpHDz:b-oly51c) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3a3e3fd01da018df742cb0793b04623381a5b2/pkg/sql/parser/overload.go#L467)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

Would you be OK with doing a typecheck on `overloadImpl`s to see if they're a `BinOp` in the filter here? I'm sort of hesitant to add more inputs to this function that already (necessarily) has a lot going on.
</blockquote></details>

The boolean here wouldn't hurt. The function still has few arguments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by justinj** [comment]
Review status: 4 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/overload.go, line 467 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-KqggLSsbpa7xDWRb5vT:-Kqwztb6W3ZKSR0HK_ik:bm3yqt3) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f3a3e3fd01da018df742cb0793b04623381a5b2/pkg/sql/parser/overload.go#L467)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

The boolean here wouldn't hurt. The function still has few arguments.
</blockquote></details>

OK, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by knz** [comment]
Reviewed 1 of 2 files at r4, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/type_check.go, line 1051 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-Kqx01neODqp23sPRHv2:-Kqx01neODqp23sPRHv3:b-cnezcq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a58c8500f4bdf433a6810f3470a97afab84da2/pkg/sql/parser/type_check.go#L1051)):*
> ```Go
> 	}
> 
> 	typedSubExprs, fns, err := typeCheckOverloadedExprs(ctx, TypeAny, ops, false, foldedLeft, foldedRight)
> ```

Can you double check whether this wouldn't want to be `true` instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by justinj** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/type_check.go, line 1051 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-Kqx01neODqp23sPRHv2:-Kqx7O1b3i526iXH020c:bcq78qa) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a58c8500f4bdf433a6810f3470a97afab84da2/pkg/sql/parser/type_check.go#L1051)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Can you double check whether this wouldn't want to be `true` instead.
</blockquote></details>

I'm not sure it makes a difference, since any comparison involving `NULL` (at least at present) will be `NULL`, unless I misunderstand what you mean by double check?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318)*

**Comment by knz** [comment]
:lgtm: modulo clarification wrt comparisons.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/type_check.go, line 1051 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-Kqx01neODqp23sPRHv2:-Kqx8Yd8hi5uqcObOmNQ:b1ae5ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a58c8500f4bdf433a6810f3470a97afab84da2/pkg/sql/parser/type_check.go#L1051)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

I'm not sure it makes a difference, since any comparison involving `NULL` (at least at present) will be `NULL`, unless I misunderstand what you mean by double check?
</blockquote></details>

Aha! I had missed that crucial detail. Maybe add this comment here:
```
// For comparisons, we do not stimulate the typing of untyped NULL with the other side's type, because
// comparisons of NULL with anything else are defined to return NULL anyways.
// Should the SQL dialect ever be extended with comparisons that can return non-NULL on NULL input, 
// the `inBinOp` parameter may need altering.
```

or something like this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-Kqx9ULQhMzVDwn76s-c:b-oo4272)*

**Comment by justinj** [comment]
Will merge on green, thanks for the reviews everyone!

---

Review status: 1 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/parser/type_check.go, line 1051 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-Kqx01neODqp23sPRHv2:-KqxB4xat4eZcnaLuxjU:b4tsatx) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5a58c8500f4bdf433a6810f3470a97afab84da2/pkg/sql/parser/type_check.go#L1051)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Aha! I had missed that crucial detail. Maybe add this comment here:
```
// For comparisons, we do not stimulate the typing of untyped NULL with the other side's type, because
// comparisons of NULL with anything else are defined to return NULL anyways.
// Should the SQL dialect ever be extended with comparisons that can return non-NULL on NULL input, 
// the `inBinOp` parameter may need altering.
```

or something like this.
</blockquote></details>

Yes, good point! I'll add a clarification to the function itself too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17318#-:-KqxBuy30HT7kr6NllfU:bmjczvf)*
